### PR TITLE
Encode Language standard in VS XML. Fixes: #6314

### PR DESCRIPTION
**Pull request by qmor**

### CODE REVIEW

**Comment by qmor** [comment]
> Please amend this PR to use a commit message of the form:
> 
> ```
> do foo when bar in order to baz
> 
> Fixes #bugid
> ```
> 
> The description of the commit needs to tell the reader what it is trying to accomplish without just saying "look at this bug report", and bug report links should be used in the postscript.
> 
> Also the exact syntax "Fixes "#bugid" will let github generate a close action when the PR is merged.

done

File: mesonbuild/backend/vs2010backend.py [hunk 1]
```diff
@@ -1099,6 +1099,11 @@ def gen_vcxproj(self, target, ofname, guid):
         else:
             ET.SubElement(clconf, 'FavorSizeOrSpeed').text = 'Speed'
         # Note: SuppressStartupBanner is /NOLOGO and is 'true' by default
+        if self.name in('vs2017', 'vs2019'):
```
Comment (dcbaker) [inline mesonbuild/backend/vs2010backend.py:1102]: `in` is a keyword, there should be a space after `in`.

File: mesonbuild/backend/vs2010backend.py [hunk 2]
```diff
@@ -1099,6 +1099,11 @@ def gen_vcxproj(self, target, ofname, guid):
...
+            if 'cpp' in file_args:
+                optargs = [x for x in file_args['cpp'] if x.startswith('/std:c++')]
+                if len(optargs)>0:
```
Comment (dcbaker) [inline mesonbuild/backend/vs2010backend.py:1105]: using `len()` to check container emptiness is a python anti-pattern, as it o(N):
```python
if optargs:
    ...
```

**Review by dcbaker — CHANGES_REQUESTED** [review]
I've got a couple of style nits, but also a question. Why is this only beind done for C++ and not for C as well?

**Comment by qmor** [comment]
> I've got a couple of style nits, but also a question. Why is this only beind done for C++ and not for C as well?

unfortunatelly i have no idea how to take right c_std option passed from meson.build file. 
i've wrote in my meson.build:

           project('test_project', ['c'], default_options : ['c_std=c11'])

and then in vs2010backend.py:

            if 'c' in file_args:
                opts = self.get_base_options_for_target(target)
                if OptionKey('std',lang='c') in opts.options:
                    optcstd = opts.options[OptionKey('std',lang='c')].value
                    if optcstd in ('c11','c17'):
                        ET.SubElement(clconf, 'LanguageStandard_C').text = 'std'+optcstd

but in some reason optcstd is always 'c99'

**Comment by dcbaker** [comment]
IIRC you need a fairly recent verison of MSVC to get c11 support, 2019 I think.

**Comment by dcbaker** [comment]
I can poke at this more tomorrow if you need some help with the C part

**Comment by qmor** [comment]
> I can poke at this more tomorrow if you need some help with the C part

Looks like it works after clean configure. Has pushed the same changes as for c++

File: mesonbuild/backend/vs2010backend.py (hunk 1 above)
Comment (qmor) [reply mesonbuild/backend/vs2010backend.py]: done

File: mesonbuild/backend/vs2010backend.py (hunk 2 above)
Comment (qmor) [reply mesonbuild/backend/vs2010backend.py]: done

**Comment by qmor** [comment]
@dcbaker, looks like it's works now with c and c++

**Comment by jpakkane** [comment]
This branch has many commits that are merges from master. These are not permitted and need to be removed. For details see:

https://mesonbuild.com/Contributing.html#keeping-pull-requests-up-to-date

**Comment by qmor** [comment]
Will open new pull request. More clear

**Pull request closed without merging**